tests/integration/integration-tests-gpu.sh

```diff
@@ -0,0 +1,271 @@
+#
+# The Admin pod is created once at the beginning of the tests and it is used to manage Spark service accounts
+# throughtout the integration tests. On the other hand, the User pod(s) are created together with the creation
+# of the Spark user (service accounts and secrets) at the beginning of each test, and they are destroyed at the
+# end of the test.
+
+
+# Import reusable utilities
+source ./tests/integration/utils/s3-utils.sh
+source ./tests/integration/utils/k8s-utils.sh
+
+# Global variables
+NAMESPACE=tests
+SERVICE_ACCOUNT=spark
+ADMIN_POD_NAME=testpod-admin
+S3_BUCKET=spark-$(uuidgen)
+
+get_spark_version(){
+  # Fetch Spark version from rockcraft.yaml
+  cat images/charmed-spark-gpu/rockcraft.yaml | yq '(.version)' 
+}
+
+
+spark_image(){
+  echo "ghcr.io/canonical/test-charmed-spark-gpu:$(get_spark_version)"
+}
+
+
+setup_user() {
+  echo "setup_user() ${1} ${2}"
+
+
+  USERNAME=$1
+  NAMESPACE=$2
+
+  kubectl -n $NAMESPACE exec testpod-admin -- env UU="$USERNAME" NN="$NAMESPACE" \
+                /bin/bash -c 'spark-client.service-account-registry create --username $UU --namespace $NN'
+
+  # Create the pod with the Spark service account
+  cat ./tests/integration/resources/testpod.yaml | yq ea '.spec.serviceAccountName = '\"${USERNAME}\"' | .spec.containers[0].image="ghcr.io/welpaolo/charmed-spark@sha256:d8273bd904bb5f74234bc0756d520115b5668e2ac4f2b65a677bfb1c27e882da"' | \
+    kubectl -n tests apply -f -
+
+  wait_for_pod testpod $NAMESPACE
+
+  TEST_POD_TEMPLATE=$(cat tests/integration/resources/podTemplate.yaml)
+  TEST_GPU_TEMPLATE=$(cat tests/integration/resources/gpu_executor_template.yaml)
+
+  kubectl -n $NAMESPACE exec testpod -- /bin/bash -c 'cp -r /opt/spark/python /var/lib/spark/'
+  kubectl -n $NAMESPACE exec testpod -- env PTEMPLATE="$TEST_POD_TEMPLATE" /bin/bash -c 'echo "$PTEMPLATE" > /etc/spark/conf/podTemplate.yaml'
+  kubectl -n $NAMESPACE exec testpod -- env PTEMPLATE="$TEST_GPU_TEMPLATE" /bin/bash -c 'echo "$PTEMPLATE" > /etc/spark/conf/gpu_executor_template.yaml'
+}
+
+setup_user_context() {
+  setup_user spark $NAMESPACE
+}
+
+cleanup_user() {
+  EXIT_CODE=$1
+  USERNAME=$2
+  NAMESPACE=$3
+
+  kubectl -n $NAMESPACE delete pod testpod --wait=true
+
+  kubectl -n $NAMESPACE exec testpod-admin -- env UU="$USERNAME" NN="$NAMESPACE" \
+                  /bin/bash -c 'spark-client.service-account-registry delete --username $UU --namespace $NN'  
+
+  OUTPUT=$(kubectl -n $NAMESPACE exec testpod-admin -- /bin/bash -c 'spark-client.service-account-registry list')
+
+  EXISTS=$(echo -e "$OUTPUT" | grep "$NAMESPACE:$USERNAME" | wc -l)
+
+  if [ "${EXISTS}" -ne "0" ]; then
+      exit 2
+  fi
+
+  if [ "${EXIT_CODE}" -ne "0" ]; then
+      exit 1
+  fi
+}
+
+cleanup_user_success() {
+  echo "cleanup_user_success()......"
+  cleanup_user 0 spark $NAMESPACE
+}
+
+cleanup_user_failure() {
+  echo "cleanup_user_failure()......"
+  cleanup_user 1 spark $NAMESPACE
+}
+
+wait_for_pod() {
```

Comment:
suggestion: `wait_for_pod` utility function is already available from `k8s-utils.sh` so the function is readily usable here.